src/filter.cc

```diff
@@ -0,0 +1,53 @@
+#include "../include/filter/filter.hh"
+
+BPFilter::BPFilter(int order, double sampling_rate, double centre_freq_, double freq_range_, Pipe<Audio> &input_,
+                   Pipe<Audio> &output_)
+    : centre_freq(centre_freq_), freq_range(freq_range_), input(input_), output(output_) {
+
+    f.setup(order, sampling_rate, centre_freq, freq_range);
+}
+BPFilter::~BPFilter() {
+    running = false;
+    if (thread_alive) {
+        filter_thread.join();
+    }
+}
+
+void BPFilter::run() {
+    filter_thread = std::thread([this]() {
+        std::unique_lock<std::mutex> lk(input.cond_m);
+        running = true;
+        thread_alive = true;
+        while (running) {
+            input.cond.wait(lk, [this] { return input.queue.empty() == false; });
+
+            Audio audio_out = filter(input.queue.front());
+            input.queue.pop();
+
```

Comment:
You should unlock your input pipe when you're not using it, unless the filter running is really quick. lk.unlock(); and lk.lock(); should do the trick. See #41 for more.

Reply:
Done

Reply:
epic